scripts/Diffraction/isis_powder/routines/common.py

```diff
@@ -707,3 +712,44 @@ def workspace_has_current(ws):
     """
     charge = ws.run().getProtonCharge()
     return charge is not None and charge > 0
+
+
+def _remove_masked_and_monitor_spectra(data_workspace: Workspace2D, correction_workspace: Workspace2D, run_details):
+    cal_workspace = mantid.LoadCalFile(
+        InputWorkspace=data_workspace,
+        CalFileName=run_details.grouping_file_path,
+        WorkspaceName="cal_workspace",
+        MakeOffsetsWorkspace=False,
+        MakeMaskWorkspace=False,
+        MakeGroupingWorkspace=True,
+    )
+
+    detectors_to_mask = []
+    for wsIndex in range(0, cal_workspace.getNumberHistograms()):
+        if cal_workspace.dataY(wsIndex) == 0:
+            detectors_to_mask.append(cal_workspace.getDetectorIDs(wsIndex)[0])
+
+    # Remove Masked and Monitor spectra
+    correction_workspace = mantid.ExtractMonitors(
```

Comment:
Could some of this be done in a loop e.g.
```
for ws in [correction_workspace, data_workspace]:
    # remove spectra
```

Reply:
The same .cal file is used to mask both workspaces so I think the spectra that get masked should be the same?

Reply:
Yep that's good, as long as we don't go through the work of getting the masked detectors twice form the same file - which i don't think it does, looks like it's done before the new loop!